### PR TITLE
Reworked Dynatrace integrations

### PR DESCRIPTION
**Pull request by aloismayr**
In this PR I've reworked the our integrations and worked on the following:
 - Renamed Ruxit to Dynatrace OneAgent
 - Renamed DynaTrace to Dynatrace Appmon
 - Updated detection criterion for Dynatrace OneAgent (SaaS/Managed)
	- tenant, tenanttoken and server are deprecated (but still supported for the sake of backward compatibilty)
	- environmentid, apitoken and endpoint are the new required fields
 - Dynatrace OneAgent will be downloaded from Dynatrace SaaS/Managed cluster via REST API (if apitoken given) rather than from the public repository
 - Public repository will still be maintained for ensuring backward compatibility
 - Updated documentation and tests
 - Minor bugfix to support caching of HTTP downloads with query parameters

### CODE REVIEW

File: lib/java_buildpack/framework/dynatrace_one_agent.rb [hunk 1]
```diff
@@ -76,26 +84,39 @@ def supports?
...
       end
 
-      def application_id
-        @application.details['application_name']
+      def agent_download_url
+        credentials = @application.services.find_service(FILTER)['credentials']
+        download_uri = server(credentials).gsub('/communication', '').gsub(':8443', '').gsub(':443', '')
+        download_uri += '/api/v1/deployment/installer/agent/unix/paas/latest?include=java&bitness=64&'
+        download_uri += "Api-Token=#{credentials[APITOKEN]}"
+        ['latest', download_uri]
```
Comment (nebhale) [inline lib/java_buildpack/framework/dynatrace_one_agent.rb]: Generally `latest` is good enough here.  Just want to be clear that not having a real version means that it'll be difficult for operators to determine which applications are vulnerable our out of date.  It's more a serviceability issue than anything else and I'd encourage you to put in a real version if possible.  If not, this is fine.

File: lib/java_buildpack/util/cache/cached_file.rb [hunk 2]
```diff
@@ -33,7 +33,7 @@ class CachedFile
         # @param [String] uri a uri which uniquely identifies the file in the cache
         # @param [Boolean] mutable whether the cached file should be mutable
         def initialize(cache_root, uri, mutable)
-          key            = URI.escape(uri.sanitize_uri, ':/')
+          key            = URI.escape(uri.sanitize_uri, ':/&')
```
Comment (nebhale) [inline lib/java_buildpack/util/cache/cached_file.rb]: Not many people have the courage to submit a change to the core functionality 😉 Definitely a needed change though.

File: config/dynatrace_appmon_agent.yml [hunk 3]
```diff
@@ -17,4 +17,4 @@
 ---
 version: 6.3.0_+
 repository_root: http://downloads.dynatracesaas.com/cloudfoundry/buildpack/java
-default_agent_name:
```
Comment (aloismayr) [inline config/dynatrace_appmon_agent.yml]: No clue how and why the space has disappeared. I added it again.

File: docs/framework-dynatrace_appmon_agent.md [hunk 4]
```diff
@@ -1,41 +1,41 @@
-# DynaTrace Agent Framework
-The DynaTrace Agent Framework causes an application to be automatically configured to work with a bound [DynaTrace Service][] instance (Free trials available).
+# Dynatrace Appmon Agent Framework
+The Dynatrace Appmon Agent Framework causes an application to be automatically configured to work with a bound [Dynatrace Service][] instance (Free trials available).
 
-The applications Cloud Foundry name is used as the `agent group` in DynaTrace, and must be pre-configured on the DynaTrace server.
+The applications Cloud Foundry name is used as the `agent group` in Dynatrace Appmon, and must be pre-configured on the Dynatrace server.
```
Comment (aloismayr) [inline docs/framework-dynatrace_appmon_agent.md]: Thank you. Changed that!

File: docs/framework-dynatrace_appmon_agent.md [hunk 5]
```diff
@@ -1,41 +1,41 @@
...
 
 **NOTE**  
 
-* The DynaTrace agent slows down app execution significantly at first, but gets faster over time. Setting the application manifest to contain `maximum_health_check_timeout` of 180 or more and/or using `cf push -t 180` or more when pushing a DynaTrace-monitored application may help.
-* Multiple `cf push`s will cause dead penguins to build up in the DynaTrace agent dashboard, as CF launches/disposes application containers. These can be hidden but will collect in the dynatrace database.
+* The Dynatrace Appmon agent slows down app execution significantly at first, but gets faster over time. Setting the application manifest to contain `maximum_health_check_timeout` of 180 or more and/or using `cf push -t 180` or more when pushing a Dynatrace-monitored application may help.
+* Multiple `cf push`s will cause dead penguins to build up in the Dynatrace Appmon dashboard, as CF launches/disposes application containers. These can be hidden but will collect in the Dynatrace database.
```
Comment (aloismayr) [inline docs/framework-dynatrace_appmon_agent.md]: Thank you. Reworked the complete NOTE section.

File: docs/framework-dynatrace_appmon_agent.md [hunk 6]
```diff
@@ -1,41 +1,41 @@
...
 
 <table>
   <tr>
-    <td><strong>Detection Criterion</strong></td><td>Existence of a single bound DynaTrace service.
+    <td><strong>Detection Criterion</strong></td><td>Existence of a single bound Dynatrace service.
```
Comment (aloismayr) [inline docs/framework-dynatrace_appmon_agent.md]: Thank you. Yeah, changed "Dynatrace service" to "Dynatrace Appmon service" where applicable and meaningful.

File: docs/framework-dynatrace_appmon_agent.md [hunk 7]
```diff
@@ -1,41 +1,41 @@
...
       <ul>
-        <li>Existence of a DynaTrace service is defined as the <a href="http://docs.cloudfoundry.org/devguide/deploy-apps/environment-variable.html#VCAP-SERVICES"><code>VCAP_SERVICES</code></a> payload containing a service who's name, label or tag has <code>dynatrace</code> as a substring and contains <code>server</code> field in the credentials. Note: The credentials must <b>NOT</b> contain <code>tenant</code> and <code>tenanttoken</code> in order to make sure the detection mechanism does not interfere with Dynatrace Ruxit integration.</li>
+        <li>Existence of a Dynatrace service is defined as the <a href="http://docs.cloudfoundry.org/devguide/deploy-apps/environment-variable.html#VCAP-SERVICES"><code>VCAP_SERVICES</code></a> payload containing a service who's name, label or tag has <code>dynatrace</code> as a substring and contains <code>server</code> field in the credentials. Note: The credentials must <b>NOT</b> contain <code>tenant</code> and <code>tenanttoken</code> in order to make sure the detection mechanism does not interfere with Dynatrace SaaS/Managed integration.</li>
```
Comment (aloismayr) [inline docs/framework-dynatrace_appmon_agent.md]: Changed "Dynatrace service" to "Dynatrace Appmon service" where applicable and meaningful.

File: docs/framework-dynatrace_appmon_agent.md [hunk 8]
```diff
@@ -1,41 +1,41 @@
...
     <td><strong>Tags</strong></td>
-    <td><tt>dyna-trace-agent=&lt;version&gt;</tt></td>
+    <td><tt>dynatrace-appmon-agent=&lt;version&gt;</tt></td>
   </tr>
 </table>
 Tags are printed to standard output by the buildpack detect script
 
 ## User-Provided Service
-Users must provide their own DynaTrace service. A user-provided DynaTrace service must have a name or tag with `dynatrace` in it so that the DynaTrace Agent Framework will automatically configure the application to work with the service.
+Users must provide their own Dynatrace service. A user-provided Dynatrace service must have a name or tag with `dynatrace` in it so that the Dynatrace Appmon Agent Framework will automatically configure the application to work with the service.
```
Comment (aloismayr) [inline docs/framework-dynatrace_appmon_agent.md]: Thanks, also changed the wording to Dynatrace Appmon service here.

File: lib/java_buildpack/framework/dynatrace_one_agent.rb (hunk 1 above)
Comment (aloismayr) [inline lib/java_buildpack/framework/dynatrace_one_agent.rb]: Agreed. At the time of compiling the download URL we don't know about the exact version we will get. We only know it is the `latest` version available for the respective environment. We could add a log-message in the `compile` or `release` step tough.

Comment (nebhale) [inline lib/java_buildpack/framework/dynatrace_one_agent.rb]: OK, no worries.  Just one of those things to think about for the longer term.  Maybe another URI that exposes the version or even a `HEAD` request to that URI which returns a header with the value in it.  Nothing that can't be added in the future.

File: lib/java_buildpack/util/cache/cached_file.rb (hunk 2 above)
Comment (aloismayr) [inline lib/java_buildpack/util/cache/cached_file.rb]: I'm sorry I don't wanted to smuggle this into the core functionality with the PR. Hope this was okay tough. Next time I'll talk to you first - or create an issue.

Comment (nebhale) [inline lib/java_buildpack/util/cache/cached_file.rb]: No, no!  This was a perfect change that I'm happy to have.  Just most people never dive deeply enough into the code to know what they need to do.  More like this please.

File: lib/java_buildpack/framework/dynatrace_one_agent.rb (hunk 1 above)
Comment (aloismayr) [inline lib/java_buildpack/framework/dynatrace_one_agent.rb]: Understand, sounds good to me.

**Pull request merged**